### PR TITLE
Improving optimize_pytorch

### PR DESCRIPTION
**Pull request by mattorourke17**
This PR does a few things:

- Adds support for the user to specify a ```device``` kwarg to the ```TNOptimizer``` constructor. This allows the user to be able to use their CPU even if they have a pytorch-compatible GPU. The previous functionality forces the user to use the GPU if they have one.

- Adds a ```TensorNetwork.multiply_each()``` method which allows the user to directly scale each tensor in the network by a given scalar.  This improves on ```multiply()``` with the ```spread_over``` kwarg in multiple ways. First, when multiplying a TensorNetwork by a scalar, typically the most numerically stable option is to spread the factor out uniformly over all the tensors (this is not the default behavior of ```multiply```). Second, in situations where the size of the tensor network is not known within the scope of a ```multiply()``` call (for an example, see the propsed changes in ```optimize_pytorch```), the ```spread_over``` kwarg is not so useful and this new function provides a more general means to appropriately rescale the network. Finally, when using ```multiply(x)``` to spread ```x``` uniformly over a network with many tensors, there is an inherent precision loss issue when computing ```x_spread = x**(1.0 / N)```. If the user has a more precise technique for computing ```x_spread```, then the new function allows them to directly apply it to every tensor.

- Adds support for a ```norm_fn_scalar``` option within the ```optimize_pytorch.TNOptimizer```. The idea behind this is explained in the documentation for the function, but the basic notion can be summarized as follows: Although the existing ```norm_fn``` option provides very general support for optimization constraints to be placed on the network that is being optimized, this means that all of the operations within ```norm_fn(tn_opt)``` are added to the computation graph and are differentiated during the optimization. For many tensor networks, we know that the normalization procedure is really just a scalar multiplication of the tensors. In this case, we just want to add that scalar multiplication to the differentiable computation graph, not all the operations the we did to obtain those scalars. Thus, if the user knows that they just want to rescale the tensors, they can use the new kwarg ```norm_fn_scalar```, which should be a function that takes the network being optimized and returns the scalar by which to multiply each tensor. Using this option, the computation of the scalars is performed outside the scope of the computation graph, which can lead to significant improvements in memory usage and cpu time when the norm computation is operation intensive (e.g. PEPS).

- Adds documentation for ```optimize_pytorch.TNOptimizer```, which seems necessary because its capabilities are now diverging a little bit from the ```optimize_tensorflow.TNOptimizer```, which is the only location for optimizer documentation.

### CODE REVIEW

**Comment by jcmgray** [comment]
Looks great, thanks! Test failure is unrelated. A few minor initial thoughts:

1. Do you know if it is possible to have ``norm_fn`` and ``norm_fn_scalar`` both at once:
```python
tn = self.norm_fn(self.tn_opt))
if self.norm_fn_scalar is not None:
    with torch.no_grad():
        fac = self.norm_fn_scalar(tn)
        tn.multiply_each_(fac)
self.loss = self.loss_fn(tn)
```
Or does turning the gradient off-and-on mid step mess things up? The specific case I have in mind is optimizing a density matrix - ``norm_fn`` you need to make the TN positive symmetric, but ``norm_fn_scalar`` you could use to actually normalize the density matrix once formed. 

2.   Might be convenient to add
```python
multiply_each_ = functools.partialmethod(multiply_each, inplace=True)
```

**Comment by mattorourke17** [comment]
The new commits support usage of both ```norm_fn``` and ```norm_fn_scalar``` in the same optimization (as well as their separate use). All of my local tests on this pass, and as far as my experience has taught me, the only major thing that pytorch complains about (for good reason) in this situation is if inplace modifications are made to the differentiable variables. As long as inplace changes are not made within the ```closure()```, then I think things like this should work.

**Comment by jcmgray** [comment]
Amazing, thanks. I *think* ``quimb`` itself never actually modifies arrays inplace (instead the ``inplace`` keyword always just refers to TN structure) so from that perspective should be okay.

Out of interest how does the performance for you work with this method? For the following example:
```python
import quimb as qu
import quimb.tensor as qtn
from quimb.tensor.optimize_pytorch import TNOptimizer

psi0 = qtn.MPS_rand_state(20, bond_dim=16)
H = qtn.MPO_ham_heis(20)

def normalize(psi):
    fac = (psi & psi).contract(all, optimize='auto-hq')
    return psi / fac**0.5

def normalize_scalar(psi):
    fac = (psi & psi).contract(all, optimize='auto-hq')
    return fac**(-0.5 / 20)

def energy(psi, H):
    energy = qtn.TensorNetwork(qtn.align_TN_1D(psi, H, psi))
    return energy.contract(all, optimize='auto-hq')

tnopt = TNOptimizer(
    tn=psi0, 
    loss_fn=energy, 
#     norm_fn=normalize,
    norm_fn_scalar=normalize_scalar,
    loss_constants={'H': H},
)

gs = tnopt.optimize(100)
```
With the standard ``norm_fn`` I get the expected ~ ``-8.6`` whereas with the scalar method it only ever reaches about half that. Possibly I am not using the new function correctly!

**Comment by mattorourke17** [comment]
I did a few tests, and I think this is related to some concerns that I had earlier, but then forgot about. 

Currently, it is not clear clear to me that any version of ```optimize_pytorch``` (this PR or the develop branch) actually controls the norm of the parameters within ```self.tn_opt```. Instead, it seems that ```self.norm_fn( self.tn_opt)``` just returns a new object, which is then supplied to ```loss_fn```, and then the optimizer figures out how to update the parameters in ```self.tn_opt``` based on this chain of operations. However, this never actually rescales the elements within ```self.tn_opt```, so the norm that ```norm_fn``` initially sees can continually get larger. If you throw some print statements into the above example you can see that the norm computed by ```norm_fn``` is huge after ~30 iterations (for some reason the norm computed by ```norm_fn_scalar``` appears to be even larger than the one computed by ```norm_fn```. Not sure why, but I'll look into it more).

I think what we really want inside the ```optimize_pytorch``` module is:

- let the optimizer take one step, updating ```self.tn_opt```
- rescale the actual parameters under optimization (```self.tn_opt```) so that the norm is close to 1
- send the "new" parameters back to the optimizer for a new step

I think this would keep the numerics under much better control. However, pytorch really does not seem to like this idea. When I try something like ```self.tn_opt = self.tn_opt.multiply_each(fac)``` inside the ```closure()```, it yells at me. And even when I try to do this outside of the ```closure```, after ```self.optimizer.step(self.closure)``` is called in ```optimize```, it still yells at me.

Do you have any idea about how to do this?

**Comment by jcmgray** [comment]
I *suspect* the problem is that ``pytorch`` traces through the input arrays to the output ``loss`` in a kind of JIT compilation, such that whilst it seems like you are interacting with it in a python like way during closure, maybe the first ``fac`` is baked into the underlying computational graph and not truly updated or something like that... but I'm not really sure, pytorch hides the details of this stuff and I will do a bit of digging.

You are right about how ``loss_fn`` and ``norm_fn`` currently work. In principle, it shouldn't matter too much if the raw arrays in ``tn_opt`` are very large (but still <1e300) as long as they are v roughly consistent from tensor to tensor (not guaranteed at the moment) and get normalized before ``loss_fn``. In other words, 3.e100 / 3.e100 is no less accurate than 3. / 3. But I agree it would feel nicer if they were actually kept updated and normalized, and doing that might also fix ``norm_fn_scalar``.

The core problem seems to be: "whats the proper way to include dynamically updating but 'non-variable' numeric parts in the computational graph of pytorch".

Did you try the ``jax`` backend out of interest? It's a bit more explicit than ``pytorch`` when it comes to what it's doing with gradients etc.

**Comment by jcmgray** [comment]
Hey @mattorourke17, hows things going on this? I've only had a chance to look very briefly into the torch side of things and to be honest things were not much clearer for me! 

But if the scalar norm function is working well enough for you then would be happy to merge this in.

**Comment by mattorourke17** [comment]
I actually have mostly figured this stuff out! I'll push the new changes asap (quite busy at the moment) and write a comment describing why it works/potential drawbacks.

**Comment by mattorourke17** [comment]
The new updates I just pushed now run your test example (and my personal code) properly. The major change was figuring out how to allow for modification of the ```tn_opt``` data during the process of optimization in order to keep the tensor network more numerically stable. It turns out that this modification of data is allowed, but not inside the ```closure```, so the rescaling according to ```norm_fn_scalar``` is shifted outside the closure into the main ```optimize``` loop. The key to make this work is that the underlying pytorch data itself has to be modified truly inplace so that the torch ```Tensors``` retain all their attributes relating to the gradients.

I didn't "improve" the call to ```norm_fn``` in the same way (i.e. allowing it to really modify the parameters that are being optimized), because it supports more general transformations and I don't see how we can guarantee that the user-supplied function will strictly respect this inplace constraint (see my comment in the code).

Other changes include:
- add more advanced stopping conditions to the optimizer based on changes in the loss function and its gradients
- change the default behavior of ```progbar=False``` so that it prints ```loss``` and ```iter time``` for each iteration, instead of just printing absolutely nothing for the entire optimization. This seems like a more reasonable default to me, but if you disagree we don't have to include this change.

**Comment by mattorourke17** [comment]
I also tried using the jax backend for some testing purposes. I agree that it's much more explicit regarding what it's doing during the backprop process, however I found that the JIT time through the gradient computation was quite long (on the order of 10+ min for my personal PEPS code), so this kind of discouraged me from further use. I'm sure I could get the time down with a bit of effort, but currently don't want to spend time on that.

I think (but this is really just a guess), that the long compile times come from jax using XLA to compile but me never utilizing any GPUs. As far as I understand XLA is a bit more of a GPU-optimized compiler and it can be known to spend a long time looking for optimizations on CPU-only code. Although this is not one of my areas of strength, so I may be completely wrong.

File: quimb/tensor/optimize_pytorch.py [hunk 1]
```diff
@@ -67,45 +72,142 @@ def parse_network_to_torch(tn, constant_tags):
 
     return tn_torch, variables, iscomplex
 
+def abs_max_grad(tn):
```
Comment (jcmgray) [inline quimb/tensor/optimize_pytorch.py]: A brief comment describing this would be good.

File: quimb/tensor/optimize_pytorch.py [hunk 2]
```diff
@@ -168,20 +278,42 @@ def optimize(self, max_steps, max_time=None):
...
+                        for tensor in self.tn_opt:
+                            tensor._data *= fac
+
                 self.optimizer.step(self.closure)
                 pbar.set_description(f"{self.loss}")
                 pbar.update()
                 self._n += 1
 
-                # check if there is a target loss we have reached
+                if self.progbar is False:
```
Comment (jcmgray) [inline quimb/tensor/optimize_pytorch.py]: I think this might be better as a separate non-default option like ``if self.progbar == 'simple'``. Having a silent option is pretty useful. 

Might also (at some later point) be worth storing the losses and times ``self.losses`` / ``self.iter_times`` etc for programmatic access rather than printing as well.

**Comment by jcmgray** [comment]
And yes just regarding JAX and XLA- the compilation time is a real negative at the moment which is a shame. I think for standard neural networks etc which have generally far fewer operations compared to tensor networks it's acceptable. Since google and others are now all interested in TNs and jax for various task I suspect it will be something people will work on. 

One perspective is that if we can keep the interface for torch, autograd, jax etc similar then experimentation can be done in whichever. Then, for final high performance runs or something 10mins is not such a bad compilation time.

On the other hand, the kind of torch specific tweak needed here shows it might be tough to keep the interface exactly the same!

**Comment by mattorourke17** [comment]
Updated according to your suggestions. 

I totally agree with your comments on JAX and trying to keep the optimizer interfaces the same. That would seem to be the ideal way to have things work on the Quimb side so that testing and production codes can both be run properly. Although one intrinsic challenge with advanced applications is to keep the user-side code completely agnostic to the auto-diff backend. For example, if a user implements a good distributed scheme to evaluate the energy, then there is a natural opportunity to also evaluate the gradient in a parallel manner. However, to do this the user needs to understand some details of the specific backend they are using and write code with a lot of backend-specific syntax in order to evaluate the gradient within their own scheme. In this sense, full support for backend-agnostic production level codes may be a challenging design problem.

**Comment by jcmgray** [comment]
Merged, thanks for the changes!
